python/lsst/daf/butler/queries/_general_query_results.py

```diff
@@ -0,0 +1,125 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("GeneralQueryResults",)
+
+from collections.abc import Iterator
+from typing import Any, final
+
+from .._dataset_ref import DatasetRef
+from .._dataset_type import DatasetType
+from ..dimensions import DataCoordinate, DimensionGroup
+from ._base import QueryResultsBase
+from .driver import QueryDriver
+from .result_specs import GeneralResultSpec
+from .tree import QueryTree, ResultColumn
+
+
+@final
+class GeneralQueryResults(QueryResultsBase):
+    """A query for `DatasetRef` results with a single dataset type.
+
+    Parameters
+    ----------
+    driver : `QueryDriver`
+        Implementation object that knows how to actually execute queries.
+    tree : `QueryTree`
+        Description of the query as a tree of joins and column expressions. The
+        instance returned directly by the `Butler._query` entry point should be
+        constructed via `make_unit_query_tree`.
+    spec : `GeneralResultSpec`
+        Specification of the query result rows, including output columns,
+        ordering, and slicing.
+
+    Notes
+    -----
+    This class should never be constructed directly by users; use `Query`
+    methods instead.
+    """
+
+    def __init__(self, driver: QueryDriver, tree: QueryTree, spec: GeneralResultSpec):
+        spec.validate_tree(tree)
+        super().__init__(driver, tree)
+        self._spec = spec
+
+    def __iter__(self) -> Iterator[dict[ResultColumn, Any]]:
+        """Iterate over result rows.
+
+        Yields
+        ------
+        row_dict : `dict` [`ResultColumn`, `Any`]
+            Result row as dictionary, the keys are `ResultColumn` instances.
```

Comment:
Another possibility for the key type here is a `<dimension-or-dataset-type-name>.<column>` string; that might be better if we make this type more publicly accessible.  In either case it should probably match however we let the user specify the desired result columns in any public interface that returns one of these.

Reply:
I switched to `str` as a key, it is indeed more natural as API uses strings as parameters. I also replaced `dataset_associations` with `general` which takes the same parameters as `GeneralResultSpec`.